linalg/petsc.cpp

```diff
@@ -480,14 +527,15 @@ PetscParVector::PetscParVector(MPI_Comm comm, const Vector &x_,
       PetscErrorCode (*rest)(Vec,PetscScalar**);
       PetscScalar *array;
 #if defined(PETSC_HAVE_DEVICE)
-      PetscBool iscuda;
-      ierr = PetscObjectTypeCompareAny((PetscObject)x,&iscuda,VECSEQCUDA,VECMPICUDA,
+      PetscBool isdevice;
+      ierr = PetscObjectTypeCompareAny((PetscObject)x,&isdevice,
+                                       VECSEQCUDA,VECMPICUDA,VECSEQHIP,VECMPIHIP,
                                        ""); PCHKERRQ(x,ierr);
-      if (iscuda && x_.UseDevice())
+      if (isdevice && x_.UseDevice())
       {
          UseDevice(true);
-         ierr = VecCUDAGetArrayWrite(x,&array); PCHKERRQ(x,ierr);
-         rest = VecCUDARestoreArrayWrite;
+         ierr = VecDeviceGetArrayWrite(x,&array); PCHKERRQ(x,ierr);
+         rest = VecDeviceRestoreArrayWrite;
```

Comment:
I wish this could just use `VecGetArrayWriteAndMemType` so it would also support kokkos vectors without listing out the types.

Reply:
I don't like the fact the `VecGetArrayWriteAndMemType` does not return the device memory when the vector is bound to the CPU. I prefer the VecCUDA/VecHIP API